pkg/storage/stores/shipper/bloomshipper/block_downloader.go

```diff
@@ -0,0 +1,224 @@
+	d.activeUsersService.UpdateUserTimestamp(tenantID, time.Now())
+	errCh := make(chan error)
+	blocksCh := make(chan blockWithQuerier, len(references))
+	go func() {
+		downloadingResultsCh := make(chan BlockDownloadingResult, len(references))
+		enqueuedCount := 0
+		downloadingParallelism := d.limits.BloomGatewayBlocksDownloadingParallelism(tenantID)
+		for _, reference := range references {
+			//todo add cache before enqueuing
+			//otherwise starts submitting the tasks to the queue and count how many tasks is submitted to know when the end reading the channel.
+			enqueuedCount++
+			task := NewBlockDownloadingTask(ctx, reference, downloadingResultsCh, errCh)
+			level.Debug(d.logger).Log("msg", "enqueuing task to download block", "block", reference.BlockPath)
+			err := d.queue.Enqueue(tenantID, nil, task, downloadingParallelism, nil)
+			if err != nil {
+				errCh <- fmt.Errorf("error enquing downloading task for block %s : %w", reference.BlockPath, err)
+				return
+			}
+		}
+		level.Debug(d.logger).Log("msg", "waiting for the downloading enqueued blocks", "count", enqueuedCount)
+		for i := 0; i < enqueuedCount; i++ {
+			select {
+			case <-ctx.Done():
+				level.Error(d.logger).Log("msg", "context is done before downloading completion", "err", ctx.Err())
+				errCh <- ctx.Err()
+				return
+			case result := <-downloadingResultsCh:
+				level.Debug(d.logger).Log("msg", "block has been downloaded and extracted", "block", result.block.BlockPath, "directory", result.extractedBlockPath)
+				blockQuerier := d.createBlockQuerier(result.extractedBlockPath)
+				blocksCh <- blockWithQuerier{
+					BlockRef:     result.block,
+					BlockQuerier: blockQuerier,
+				}
+			}
+		}
+	}()
+	return blocksCh, errCh
+}
+
+type blockWithQuerier struct {
+	BlockRef
+	*v1.BlockQuerier
+}
+
+// extract the files into directory and returns absolute path to this directory.
+func (d *blockDownloader) extractBlock(block *Block, ts time.Time) (string, error) {
+	workingDirectoryPath := filepath.Join(d.workingDirectory, block.BlockPath, strconv.FormatInt(ts.UnixMilli(), 10))
+	err := os.MkdirAll(workingDirectoryPath, os.ModePerm)
+	if err != nil {
+		return "", fmt.Errorf("can not create directory to extract the block: %w", err)
+	}
+	archivePath, err := writeDataToTempFile(workingDirectoryPath, block)
+	if err != nil {
+		return "", fmt.Errorf("error writing data to temp file: %w", err)
+	}
+	defer func() {
+		os.Remove(archivePath)
+		// todo log err
+	}()
+	err = extractArchive(archivePath, workingDirectoryPath)
+	if err != nil {
+		return "", fmt.Errorf("error extracting archive: %w", err)
+	}
+	return workingDirectoryPath, nil
+}
+
+func (s *blockDownloader) createBlockQuerier(directory string) *v1.BlockQuerier {
+	reader := v1.NewDirectoryBlockReader(directory)
+	block := v1.NewBlock(reader)
+	return v1.NewBlockQuerier(block)
+}
+
+func writeDataToTempFile(workingDirectoryPath string, block *Block) (string, error) {
```

Comment:
Maybe add a comment that this function is supposed to close the block.

Reply:
done